app/models/gem_typo.rb

```diff
@@ -0,0 +1,41 @@
+require "rubygems/text"
+
+class GemTypo
+  attr_reader :protected_gem
+
+  include Gem::Text
+
+  DOWNLOADS_THRESHOLD = 10_000_000
+  SIZE_THRESHOLD = 4
+
+  def initialize(rubygem_name)
+    @rubygem_name = rubygem_name.downcase
+  end
+
+  def protected_typo?
+    return false if @rubygem_name.size < GemTypo::SIZE_THRESHOLD
+
+    protected_gems.each do |protected_gem|
+      distance = levenshtein_distance(@rubygem_name, protected_gem)
+      if distance <= distance_threshold
+        @protected_gem = protected_gem
+        return true
+      end
+    end
+
+    false
+  end
+
+  private
+
+  def distance_threshold
+    @rubygem_name.size == GemTypo::SIZE_THRESHOLD ? 1 : 2
```

Comment:
What about to "memoize" this to not execute same call on every iteration?

Reply:
Moved it to the initializer.